students/tri_nguyen/lesson03/list_lab.py

```diff
@@ -0,0 +1,179 @@
+for fruit in fruit_list:
+    if fruit.startswith('P'):
+        print(fruit)
+
+# End of Series 1
+# ------------------------------------------- #
+
+# ------------------------------------------- #
+# Series 2 - Using the list created in series 1:
+# - Display the list.
+# - Remove the last fruit from the list.
+# - Display the list.
+# - Ask the user for a fruit to delete, find it
+#   and delete it.
+# - (Bonus: Multiply the list times two. Keep asking
+#    until a match is found. Once found, delete all
+#    occurrences).
+
+print('\n'+'--Series 2 of the exercise--'.upper()+'\n')
+
+print('Below is the current list of fruits.')
+for idx, fruit in enumerate(fruit_list):
+    print(idx + 1, '-', fruit)
+
+print()
+print('The last fruit ({}) from the list has been removed.'.format(fruit_list[len(fruit_list) - 1]))
+fruit_list.pop()
+for idx, fruit in enumerate(fruit_list):
+    print(idx + 1, '-', fruit)
+
+print()
+to_be_deleted = input('Enter a fruit to delete: ')
+
+if to_be_deleted.title() in fruit_list:
+    fruit_list.remove(to_be_deleted.title())
+
+print('\nBelow is a newly updated list of fruits.')
+for idx, fruit in enumerate(fruit_list):
+    print(idx + 1, '-', fruit)
+
+
+# End of Series 2
+# ------------------------------------------- #
+
+# ------------------------------------------- #
+# Series 3 - Using the list from series 1:
+# - Ask the user for input displaying a line like
+#   'Do you like apples?' for each fruit in the list.
+#   make the fruit all lowercase.
+# - For each 'no', delete that fruit from the list.
+# - For any answer that is not 'yes' or 'no', prompt
+#   the user to answer with one of those two values.
+# - Display the list.
+
+print('\n'+'--Series 3 of the exercise--'.upper()+'\n')
+
+count = 0
+answer_set = ('yes', 'no')
+
+while count <= len(fruit_list):
+
+    for idx, fruit in enumerate(fruit_list):
+
+        from_user = input('Do you like ' + fruit.lower() + '? ')
+
+        if from_user.lower() == answer_set[1]:
+            print("\nYou don't like " + fruit.lower() + ". It'll be removed.\n")
+            fruit_list.remove(fruit)
+        elif from_user.lower() == answer_set[0]:
+            print('\n' + fruit + ' are good.\n')
+        else:
+            print('\nPlease enter yes or no\n')
+            continue
+        count = count + 1
+
+print()
+print('Below is the newly updated list of fruits.')
+for idx, fruit in enumerate(fruit_list):
+    print(idx + 1, '-', fruit)
+
+# End of Series 3
+# ------------------------------------------- #
+
+# ------------------------------------------- #
+# Series 4 - Using the list from series 1
+# - Make a copy of the list and reverse the letters
+#   in each fruit in the copy.
+# - Delete the last item of the original list.
+# - Display the original and the copy.
+# ------------------------------------------- #
+
+print('\n'+'--Series 4 of the exercise--'.upper()+'\n')
+
+copied_fruit_list = fruit_list.copy()
```

Comment:
even though using `copy()` here works,  warning on copying of objects: 

I would say most of the time you should use deep copy to avoid surprises where your original data structure gets modified. I often use `deepcopy` method because it can apply to lists/dicts:
```
from copy import deepcopy
list_copy = deepcopy(original_list)
```
you can also copy list this way: `original_list[:]`

-----
here is an example of why `copy()` (shallow copy) can do some unexpected things:

```
>>> fruit_list_complex = [['a', 'b'], ['c']]
>>> fruit_list_complex_copy = fruit_list_complex.copy()
>>> fruit_list_complex
[['a', 'b'], ['c']]
>>> fruit_list_complex_copy
[['a', 'b'], ['c']]
>>> fruit_list_complex[1].append('d')
>>> fruit_list_complex
[['a', 'b'], ['c', 'd']]
>>> fruit_list_complex_copy
[['a', 'b'], ['c', 'd']]
>>>
```
----

the top answer [here](https://stackoverflow.com/questions/3975376/understanding-dict-copy-shallow-or-deep) does a great job explaining this in good detail.

Reply:
Wonderful tips and advice. I will clean up the exercises and resubmit.
I was not sure whether I would have to closely follow procedures/instructions from the assignments. I was going to create functions, but the assignment didn't mention so I decided not to. Going forward, I will create functions as much as I can.

Reply:
understandable; you can use functions in pretty much any of the assignments, and certainly not required for early assignment but a great way to organize the code better.